### PR TITLE
test: 100% coverage of My expense

### PR DESCRIPTION
**Pull request by suyashpatil78**
### <samp>🤖 Generated by Copilot at 4272fcd</samp>

This pull request enhances the user interface and the code quality of the `MyExpensesPage` component. It imports a utility function, adds a null check, removes an unnecessary operator, and creates mock data for the snackbar messages.

### <samp>🤖 Generated by Copilot at 4272fcd</samp>

> _Oh we're the crew of the `MyExpensesPage`_
> _We test and code with skill and rage_
> _We mock the snackbar and the report_
> _And delete the expenses with a null abort_

### <samp>🤖 Generated by Copilot at 4272fcd</samp>

*  Add mock snackbar properties for testing expense deletion ([link](https://github.com/fylein/fyle-mobile-app/pull/2144/files?diff=unified&w=0#diff-fce5303a5a2a7588359d78dab7c75735cc2804b23850c4bec93de211012f5437R376-R393))
*  Import lodash `assign` function to merge report and expense objects ([link](https://github.com/fylein/fyle-mobile-app/pull/2144/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L42-R42))
*  Initialize `offlineExpenses` array to store unsynced expenses ([link](https://github.com/fylein/fyle-mobile-app/pull/2144/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1274-R1274))
*  Check for null `pendingTransactions` before selecting all expenses ([link](https://github.com/fylein/fyle-mobile-app/pull/2144/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1350-R1350))
*  Remove optional chaining operator from `selectedElements` array ([link](https://github.com/fylein/fyle-mobile-app/pull/2144/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1373-R1378))

## Clickup
https://app.clickup.com/t/85zt7rr5t

## Code Coverage
Please add code coverage here

## UI Preview
Please add screenshots for UI changes

### CODE REVIEW

**Comment by suyashpatil78** [comment]
Screenshot 2023-06-22 at 3 08 59 PM

File: src/app/fyle/my-expenses/my-expenses.page.spec.ts [hunk 1]
```diff
@@ -2220,40 +2232,421 @@ describe('MyExpensesPage', () => {
...
+      spyOn(component, 'showAddToReportSuccessToast');
+    });
+    it('should call matBottomSheet.open and call showAddToReportSuccessToast if data.report is defined', () => {
+      spyOn(component, 'addTransactionsToReport').and.returnValue(of(apiExtendedReportRes[0]));
+      matBottomsheet.open.and.returnValue({
+        afterDismissed: () =>
+          of({
+            report: apiExtendedReportRes[0],
+          }),
+      } as MatBottomSheetRef<ExtendedReport>);
```
Comment (yksolanki9) [inline src/app/fyle/my-expenses/my-expenses.page.spec.ts:2347]: Type assertion not needed here. `open()` is generic

File: src/app/fyle/my-expenses/my-expenses.page.spec.ts [hunk 2]
```diff
@@ -2220,40 +2232,421 @@ describe('MyExpensesPage', () => {
...
+
+      component.showOldReportsMatBottomSheet();
+
+      expect(transactionService.getReportableExpenses).toHaveBeenCalledOnceWith(apiExpenseRes);
+      expect(matBottomsheet.open).toHaveBeenCalledOnceWith(<any>AddTxnToReportDialogComponent, {
```
Comment (yksolanki9) [inline src/app/fyle/my-expenses/my-expenses.page.spec.ts:2352]: no type assertions

File: src/app/fyle/my-expenses/my-expenses.page.spec.ts [hunk 3]
```diff
@@ -2220,40 +2232,421 @@ describe('MyExpensesPage', () => {
...
+      const mockReportData = cloneDeep(apiExtendedReportRes);
+      mockReportData[0].rp_state = 'DRAFT';
+      component.openReports$ = of(mockReportData);
+      spyOn(component, 'addTransactionsToReport').and.returnValue(of(mockReportData[0]));
+      matBottomsheet.open.and.returnValue({
+        afterDismissed: () =>
+          of({
+            report: mockReportData[0],
+          }),
+      } as MatBottomSheetRef<ExtendedReport>);
```
Comment (yksolanki9) [inline src/app/fyle/my-expenses/my-expenses.page.spec.ts:2372]: same comments everywhere

File: src/app/fyle/my-expenses/my-expenses.page.spec.ts [hunk 4]
```diff
@@ -2220,40 +2232,421 @@ describe('MyExpensesPage', () => {
...
+        panelClass: ['mat-bottom-sheet-1'],
+      });
+
+      expect(component.addTransactionsToReport).toHaveBeenCalledOnceWith(mockReportData[0], ['tx3nHShG60zq']);
+      expect(component.showAddToReportSuccessToast).toHaveBeenCalledOnceWith({
+        message: 'Expenses added to an existing draft report',
+        report: mockReportData[0],
+      });
+    });
+    it('should call matBottomSheet.open and should not call showAddToReportSuccessToast if data.report is null', () => {
```
Comment (yksolanki9) [inline src/app/fyle/my-expenses/my-expenses.page.spec.ts:2387]: empty line between test cases. Repeated comment from previous pr

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 5]
```diff
@@ -1270,8 +1271,21 @@ export class MyExpensesPage implements OnInit {
     await actionSheet.present();
   }
 
+  selectedExpensesPopoverDeleteMethod(): Observable<Transaction[]> {
+    const offlineExpenses = this.expensesToBeDeleted?.filter((expense) => !expense.tx_id);
```
Comment (yksolanki9) [inline src/app/fyle/my-expenses/my-expenses.page.ts:1275]: why ?.

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 6]
```diff
@@ -1270,8 +1271,21 @@ export class MyExpensesPage implements OnInit {
     await actionSheet.present();
   }
 
+  selectedExpensesPopoverDeleteMethod(): Observable<Transaction[]> {
```
Comment (yksolanki9) [inline src/app/fyle/my-expenses/my-expenses.page.ts:1274]: Rename to deleteSelectedExpenses

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 7]
```diff
@@ -1270,8 +1271,21 @@ export class MyExpensesPage implements OnInit {
...
+
+    this.selectedElements = this.expensesToBeDeleted?.filter((expense) => expense.tx_id);
+    if (this.selectedElements?.length > 0) {
+      return this.transactionService.deleteBulk(this.selectedElements.map((selectedExpense) => selectedExpense.tx_id));
+    } else {
+      return of(null);
+    }
+  }
+
   async deleteSelectedExpenses() {
```
Comment (yksolanki9) [inline src/app/fyle/my-expenses/my-expenses.page.ts:1287]: Rename to openDeleteExpensesPopover

File: src/app/fyle/my-expenses/my-expenses.page.ts (hunk 5 above)
Comment (yksolanki9) [inline src/app/fyle/my-expenses/my-expenses.page.ts:1275]: pass offlineExpenses as an argument here. This way we can avoid two traversals

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 8]
```diff
@@ -1291,20 +1305,7 @@ export class MyExpensesPage implements OnInit {
...
-          this.selectedElements = this.expensesToBeDeleted.filter((expense) => expense.tx_id);
-          if (this.selectedElements?.length > 0) {
-            return this.transactionService.deleteBulk(
-              this.selectedElements.map((selectedExpense) => selectedExpense.tx_id)
-            );
-          } else {
-            return of(null);
-          }
-        },
+        deleteMethod: this.selectedExpensesPopoverDeleteMethod.bind(this),
```
Comment (yksolanki9) [inline src/app/fyle/my-expenses/my-expenses.page.ts:1308]: use arrow function

File: src/app/fyle/my-expenses/my-expenses.page.spec.ts [hunk 9]
```diff
@@ -2220,40 +2232,421 @@ describe('MyExpensesPage', () => {
...
+      );
+      transactionService.getDeleteDialogBody.and.returnValue('Once deleted, the action cannot be undone');
+      component.expensesToBeDeleted = apiExpenseRes;
+      component.cccExpenses = 1;
+      transactionService.deleteBulk.and.returnValue(of(txnList));
+      snackbarProperties.setSnackbarProperties.and.returnValue(snackbarPropertiesRes3);
+      spyOn(component, 'doRefresh');
+    });
+
+    it('should open a popover and extract on dismiss', fakeAsync(() => {
```
Comment (yksolanki9) [inline src/app/fyle/my-expenses/my-expenses.page.spec.ts:2460]: extract?

File: src/app/fyle/my-expenses/my-expenses.page.spec.ts [hunk 10]
```diff
@@ -2220,40 +2232,421 @@ describe('MyExpensesPage', () => {
...
+        componentProps: {
+          header: 'Delete Expense',
+          body: 'Once deleted, the action cannot be undone',
+          ctaText: 'Exclude and Delete',
+          disableDelete: false,
+          deleteMethod: jasmine.any(Function),
+        },
+      });
+    }));
+    it('should open a popover and extract on dismiss if expenseToBeDeleted is undefined', fakeAsync(() => {
```
Comment (yksolanki9) [inline src/app/fyle/my-expenses/my-expenses.page.spec.ts:2484]: Delete cta should be disabled if expenseToBeDeleted is undefined, fix this in html

Comment (yksolanki9) [reply src/app/fyle/my-expenses/my-expenses.page.spec.ts:2484]: Else why should we open the popover if no expenses are to be deleted

File: src/app/fyle/my-expenses/my-expenses.page.spec.ts [hunk 11]
```diff
@@ -2220,40 +2232,421 @@ describe('MyExpensesPage', () => {
...
+      component.isReportableExpensesSelected = false;
+      component.onSelectAll(false);
+      expect(component.selectedElements).toEqual([]);
+      expect(transactionService.getReportableExpenses).toHaveBeenCalledOnceWith([]);
+      expect(component.isReportableExpensesSelected).toBeTrue();
+      expect(component.setExpenseStatsOnSelect).toHaveBeenCalledTimes(1);
+    });
+    it('should update selectedElements, allExpensesCount and call apiV2Service', () => {
+      transactionService.getAllExpenses.and.returnValue(of(cloneDeep(expenseList4)));
+      component.loadData$ = new BehaviorSubject({ pageNumber: 1 });
```
Comment (yksolanki9) [inline src/app/fyle/my-expenses/my-expenses.page.spec.ts:2608]: Move inside beforeEach

File: src/app/fyle/my-expenses/my-expenses.page.spec.ts [hunk 12]
```diff
@@ -2220,40 +2232,421 @@ describe('MyExpensesPage', () => {
...
+      });
+      expect(transactionService.excludeCCCExpenses).toHaveBeenCalledOnceWith([...apiExpenseRes, ...expenseList4]);
+      expect(transactionService.getDeletableTxns).toHaveBeenCalledOnceWith([...apiExpenseRes, ...expenseList4]);
+      expect(component.cccExpenses).toBe(3);
+      expect(component.selectedElements).toEqual([...apiExpenseRes, ...expenseList4]);
+      expect(component.allExpensesCount).toBe(4);
+      expect(component.isReportableExpensesSelected).toBeTrue();
+      expect(component.setExpenseStatsOnSelect).toHaveBeenCalledTimes(2);
+    });
+    it('should update selectedElements, allExpensesCount and call apiV2Service if getDeletableTxns, excludeCCCExpenses and pendingTransactions are undefined', () => {
```
Comment (yksolanki9) [inline src/app/fyle/my-expenses/my-expenses.page.spec.ts:2627]: Can you explain when this case can occur? We're setting `selectedElements` to [] at the start of this method, so I dont see any way expensesToBeDeleted can be undefined

**Review by yksolanki9 — CHANGES_REQUESTED** [review]

File: src/app/fyle/my-expenses/my-expenses.page.spec.ts (hunk 1 above)
Comment (suyashpatil78) [reply src/app/fyle/my-expenses/my-expenses.page.spec.ts:2347]: That is throwing an error without assertion

File: src/app/fyle/my-expenses/my-expenses.page.spec.ts [hunk 13]
```diff
@@ -2220,40 +2232,421 @@ describe('MyExpensesPage', () => {
...
+
+  it('addTransactionsToReport(): should show loader call reportService and hide the loader', (done) => {
+    loaderService.showLoader.and.resolveTo();
+    loaderService.hideLoader.and.resolveTo(true);
+
+    reportService.addTransactions.and.returnValue(of(apiExtendedReportRes));
+    component
+      .addTransactionsToReport(apiExtendedReportRes[0], ['tx5fBcPBAxLv'])
+      .pipe(
+        tap((updatedReport) => {
```
[the remark this answers is not in the document]
Comment (suyashpatil78) [reply src/app/fyle/my-expenses/my-expenses.page.spec.ts:2319]: I have already tried it inside subscribe block but the finalize block was not getting executed. When discussing with @Julias0, we reached upon this solution. In `auth.service.ts` we have used same thing: https://github.com/fylein/fyle-mobile-app/blob/23ffab583bb7fefeea896e056a826057290fffaf/src/app/core/services/auth.service.spec.ts#L153

File: src/app/fyle/my-expenses/my-expenses.page.ts (hunk 5 above)
Comment (suyashpatil78) [reply src/app/fyle/my-expenses/my-expenses.page.ts:1275]: As `ctaText` and `disableDelete` has optional chaining for expensesToBeDeleted, therefore kept here as well.

Comment (suyashpatil78) [reply src/app/fyle/my-expenses/my-expenses.page.ts:1275]: We don't need `?.`, removing it

File: src/app/fyle/my-expenses/my-expenses.page.spec.ts (hunk 12 above)
Comment (suyashpatil78) [reply src/app/fyle/my-expenses/my-expenses.page.spec.ts:2627]: Well, in `selectExpense` we have similar way of setting `expensesToBeDeleted`. The `getDeletableTxns` and `excludeCCCExpenses` method will return an array, so there expensesToBeDeleted cannot be undefined. removing it

File: src/app/fyle/my-expenses/my-expenses.page.ts (hunk 8 above)
Comment (suyashpatil78) [reply src/app/fyle/my-expenses/my-expenses.page.ts:1308]: On using arrow function, `deleteMethod: () => this.deleteSelectedExpenses(offlineExpenses),` statement not getting coverage

File: src/app/fyle/my-expenses/my-expenses.page.spec.ts [hunk 14]
```diff
@@ -2430,29 +2399,29 @@ describe('MyExpensesPage', () => {
...
       expect(component.selectedElements).toEqual(expenseList4);
       expect(transactionService.deleteBulk).toHaveBeenCalledOnceWith(['txKFqMRPNLsa', 'txc5zbIpTGMU', 'txo3tuIb7em4']);
     });
-    it('should update selectedElements and should not invoke deleteBulk method if expenseToBeDeleted is undefined', () => {
-      component.pendingTransactions = [];
-      component.expensesToBeDeleted = undefined;
-      component.selectedExpensesPopoverDeleteMethod();
-      expect(transactionOutboxService.deleteBulkOfflineExpenses).toHaveBeenCalledOnceWith([], undefined);
-      expect(component.selectedElements).toEqual(undefined);
+    it('should not call deleteBulk method if tx_id is not present in expensesToBeDeleted', () => {
```
Comment (yksolanki9) [inline src/app/fyle/my-expenses/my-expenses.page.spec.ts:2413]: Empty lines are still not added

File: src/app/fyle/my-expenses/my-expenses.page.spec.ts [hunk 15]
```diff
@@ -2220,40 +2232,421 @@ describe('MyExpensesPage', () => {
...
+          expect(loaderService.showLoader).toHaveBeenCalledOnceWith('Adding transaction to report');
+          expect(reportService.addTransactions).toHaveBeenCalledOnceWith('rprAfNrce73O', ['tx5fBcPBAxLv']);
+          expect(updatedReport).toEqual(apiExtendedReportRes[0]);
+        }),
+        finalize(() => {
+          expect(loaderService.hideLoader).toHaveBeenCalledTimes(1);
+        })
+      )
+      .subscribe(noop);
+    done();
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply src/app/fyle/my-expenses/my-expenses.page.spec.ts:2304]: not resolved

File: src/app/fyle/my-expenses/my-expenses.page.ts (hunk 8 above)
Comment (yksolanki9) [reply src/app/fyle/my-expenses/my-expenses.page.ts]: You can create a separate method for this. Check with @CODEX108, she has handled this.
You shouldnt be compromising on latest practices just for coverage

**Review by yksolanki9 — APPROVED** [review]
Approving the pr, but some comments are still unresolved

**Pull request merged**